README.md
# react-form-stepper

![version](https://img.shields.io/npm/v/react-form-stepper?color=brightgreen)
![GitHub Workflow Status (branch)](https://img.shields.io/github/workflow/status/M0kY/react-form-stepper/Release/master)
![David](https://img.shields.io/david/M0kY/react-form-stepper)
![David](https://img.shields.io/david/dev/M0kY/react-form-stepper)
![License](https://img.shields.io/npm/l/react-form-stepper?color=brightgreen)
![Downloads](https://img.shields.io/npm/dm/react-form-stepper)

[![NPM](https://nodei.co/npm/react-form-stepper.png?downloads=true&stars=true)](https://nodei.co/npm/react-form-stepper/)

A simple react stepper component for multi-step forms inspired by the Stepper component from Material-UI.

![react-form-stepper preview](https://raw.githubusercontent.com/M0kY/react-form-stepper/master/react-form-stepper-demo.png)

## Examples

For examples of react-form-stepper go to: https://m0ky.github.io/react-form-stepper/.

## Getting started

Install the library by running:

```sh
npm install react-form-stepper --save
```

or

```sh
yarn add react-form-stepper
```

## Usage

```js
import { Stepper } from 'react-form-stepper';
```

There are 2 possible ways of defining the steps in the Stepper component:

- Using the `steps` prop
- Using the `Stepper` as a HOC with `Step` as children

### Using the steps prop

```js
<Stepper
  steps={[{ label: 'Step 1' }, { label: 'Step 2' }, { label: 'Step 3' }]}
  activeStep={2}
/>
```

### Using the Stepper as a HOC with Step as children

```js
<Stepper activeStep={1}>
  <Step label="Children Step 1" />
  <Step label="Children Step 2" />
  <Step label="Children Step 3" />
</Stepper>
```

### Using with SSR
When developing an SSR application with a framework like `Next.js` you might face your console being polluted with the following message
`Warning: [JSS] Rule is not linked. Missing sheet option "link: true".` caused by the underlying dependency `react-jss`.
A workaround is to use the `dynamic` import module like in the example below.

```js
// CustomStepper.js
const CustomStepper = () => {
  return <Stepper steps={[{ label: 'Step 1' }, { label: 'Step 2' }]} activeStep={1} />;
};

export default CustomStepper;
```

```js
// MultiStepForm.js
import dynamic from 'next/dynamic';

const StepperComponent = dynamic(() => import('./CustomStepper'), {
  ssr: false,
});
```

### Stepper props

| Props                | Options              | Default | Description                                                            |
| -------------------- | -------------------- | ------- | ---------------------------------------------------------------------- |
| steps                | \[StepsDTO\]         | none    | Array of objecst defining the steps                                    |
| activeStep           | number               | 0       | Value defining the active step                                         |
| connectorStateColors | boolean              | false   | Use different colors for connector lines based on adjacent steps state |
| className            | string               | none    | Add css classes to the Stepper component                               |
| stepClassName        | string               | none    | Add css classes to Step components                                     |
| hideConnectors       | boolean \|'inactive' | false   | Value defining connectors visibility                                   |
| nonLinear            | boolean              | false   | Allow users to enter the flow at any point                             |
| styleConfig          | StepStyleDTO         | ---     | Object containing Step style defaults                                  |
| connectorStyleConfig | ConnectorStyleProps  | ---     | Object containing Connector style defaults                             |

### ConnectorStyleProps

| Props          | Options         | Default   | Description                                                                                                     |
| -------------- | --------------- | --------- | --------------------------------------------------------------------------------------------------------------- |
| disabledColor  | string          | '#bdbdbd' | Define the disabled connector line color                                                                        |
| activeColor    | string          | '#ed1d24' | Define the active connector line color                                                                          |
| completedColor | string          | '#a10308' | Define the completed connector line color                                                                       |
| size           | React.ReactText | 1         | Define the thickness of the connector line                                                                      |
| stepSize       | React.ReactText | '2em'     | Value defaulting to the step size, used to calculate the padded space between the step and connector line start |
| style          | string          | 'solid'   | Define the style of the connector line                                                                          |

### StepsDTO

| Props     | Options | Default | Description                                                 |
| --------- | ------- | ------- | ----------------------------------------------------------- |
| label     | string  | ''      | Value to be displayed under each step                       |
| active    | boolean | false   | Value to indicate should the step be displayed as active    |
| completed | boolean | false   | Value to indicate should the step be displayed as completed |

### StepStyleDTO

| Props              | Options          | Default    | Description                                             |
| ------------------ | ---------------- | ---------- | ------------------------------------------------------- |
| activeBgColor      | string           | '#ed1d24'  | Define the background color for active steps            |
| activeTextColor    | string           | '#ffffff'  | Define the text color for active steps                  |
| completedBgColor   | string           | '#a10308'  | Define the background color for completed steps         |
| completedTextColor | string           | '#ffffff'  | Define the background color for completed steps         |
| inactiveBgColor    | string           | '#e0e0e0'  | Define the background color for inactive steps          |
| inactiveTextColor  | string           | '#ffffff'  | Define the background color for inactive steps          |
| size               | string or number | '2em'      | Value representing the `width` and `height` of the step |
| circleFontSize     | string or number | '1rem'     | Font size of the step content                           |
| labelFontSize      | string or number | '0.875rem' | Font size of step labels                                |
| borderRadius       | string or number | '50%'      | Step border radius                                      |
| fontWeight         | string or number | 500        | Step label font weight                                  |

### Step props

| Props     | Options         | Default    | Description                                                 |
| --------- | --------------- | ---------- | ----------------------------------------------------------- |
| label     | string          | ''         | Value to be displayed under each step                       |
| active    | boolean         | false      | Value to indicate should the step be displayed as active    |
| completed | boolean         | false      | Value to indicate should the step be displayed as completed |
| index     | number          | 0          | Index value of the step                                     |
| className | string          | none       | Add css classes to the Step component                       |
| children  | React.ReactNode | Step index | Value inside the step                                       |


CHANGELOG.md
## [2.0.3](https://github.com/M0kY/react-form-stepper/compare/v2.0.2...v2.0.3) (2022-05-31)


### Bug Fixes

* override for stepSize in connectorStyleConfig ([d6ab466](https://github.com/M0kY/react-form-stepper/commit/d6ab4668fe5f76062d3fa53c5c09f1334ec4d625))

## [2.0.2](https://github.com/M0kY/react-form-stepper/compare/v2.0.1...v2.0.2) (2022-05-31)


### Bug Fixes

* react-jss dynamic props in react 18 ([185ffa5](https://github.com/M0kY/react-form-stepper/commit/185ffa5faa514e9d453b3eda026ee9a202a1f46f))

## [2.0.1](https://github.com/M0kY/react-form-stepper/compare/v2.0.0...v2.0.1) (2022-03-20)


### Bug Fixes

* add missing id to step container HTML tag ([66da326](https://github.com/M0kY/react-form-stepper/commit/66da326feb739f24f6be2b1a7bc88848eeece514))

# [2.0.0](https://github.com/M0kY/react-form-stepper/compare/v1.4.3...v2.0.0) (2022-03-20)


### Bug Fixes

* custom class name generator disable to close [#387](https://github.com/M0kY/react-form-stepper/issues/387) ([a86ae66](https://github.com/M0kY/react-form-stepper/commit/a86ae66e3fc8da1e31fece06e3ecf75e029787fa))


### BREAKING CHANGES

* generated class names change and added IDs on
HTML elements equal to previous class name values

## [1.4.3](https://github.com/M0kY/react-form-stepper/compare/v1.4.2...v1.4.3) (2021-07-23)


### Bug Fixes

* downgrade color due to error in gh-pages build ([2026669](https://github.com/M0kY/react-form-stepper/commit/20266692deb0dc410ea93d4f0ec022a21ab2c1f3))

## [1.4.2](https://github.com/M0kY/react-form-stepper/compare/v1.4.1...v1.4.2) (2021-07-23)


### Bug Fixes

* active, completed, disabled condition checks to enable overriding ([3adfb24](https://github.com/M0kY/react-form-stepper/commit/3adfb24925de30269f3d7ad4dbb9233b62074043))
* add umd build ([7ca7d53](https://github.com/M0kY/react-form-stepper/commit/7ca7d535532b8275a6c28fcd23441c3ffbd2e0a6))
* consolidate multiple css classes into a single one ([f5770ef](https://github.com/M0kY/react-form-stepper/commit/f5770efadde0ba77e25bb6a2e38f9871d6d9ea9b))
* import connector from index.ts ([eee59e1](https://github.com/M0kY/react-form-stepper/commit/eee59e113b61f69eaecd66163fcc53abfc22c317))
* make activeStep prop optional, if not defined defaults to 0 ([2867fb8](https://github.com/M0kY/react-form-stepper/commit/2867fb8c594b7de655d1b67f9370c74b646886fb))
* make StepButon prop contentClasses optional ([0dcdea1](https://github.com/M0kY/react-form-stepper/commit/0dcdea14d8291575bc4ec2a8413a1b62e9a27227))
* StepLabel style types ([8e8a8f9](https://github.com/M0kY/react-form-stepper/commit/8e8a8f97d91e3af6b7647b46744f062cdbc349dc))

## [1.4.1](https://github.com/M0kY/react-form-stepper/compare/v1.4.0...v1.4.1) (2021-04-18)


### Bug Fixes

* revert @semantic-release/npm upgrade due to breaking gh-pages:build script ([6970e75](https://github.com/M0kY/react-form-stepper/commit/6970e7545e71a3fef789762dde1d1954d14c87f1))

# [1.4.0](https://github.com/M0kY/react-form-stepper/compare/v1.3.0...v1.4.0) (2021-03-17)


### Bug Fixes

* connector style props interface ([d44b581](https://github.com/M0kY/react-form-stepper/commit/d44b58130d5638e0d386996a48066baf4895bd1b))
* revert default style colors ([7d5a547](https://github.com/M0kY/react-form-stepper/commit/7d5a5473dd0df0dc7e94c9f584d2a3619febc48e))


### Features

* add hover and active background color modifiers ([18ac984](https://github.com/M0kY/react-form-stepper/commit/18ac98435107bcb978a48f5e112ffaa46b1711c1))

# [1.3.0](https://github.com/M0kY/react-form-stepper/compare/v1.2.11...v1.3.0) (2020-12-24)


### Features

* add option to allow users to enter the flow at any point ([cda7907](https://github.com/M0kY/react-form-stepper/commit/cda7907ac23b042920ab703bad84265978374db6))
* add option to use any div element prop on Stepper component ([697f347](https://github.com/M0kY/react-form-stepper/commit/697f347477b2be19ee32e6d06aa9ba55684fa2f8))

## [1.2.11](https://github.com/M0kY/react-form-stepper/compare/v1.2.10...v1.2.11) (2020-11-28)


### Bug Fixes

* **types:** step props interface extends button element interface ([ff05f67](https://github.com/M0kY/react-form-stepper/commit/ff05f6723575664f4a7a2fa8ac0e0289fa05ec59))

## [1.2.10](https://github.com/M0kY/react-form-stepper/compare/v1.2.9...v1.2.10) (2020-06-29)


### Bug Fixes

* font size calculation condition ([c28a3c0](https://github.com/M0kY/react-form-stepper/commit/c28a3c068d3772db6c5dc43a7ed4fb30e96404a3))

## [1.2.9](https://github.com/M0kY/react-form-stepper/compare/v1.2.8...v1.2.9) (2020-06-29)


### Bug Fixes

* release content ([743698b](https://github.com/M0kY/react-form-stepper/commit/743698ba8838041dce0df07716ac908bcd560058))

## [1.2.8](https://github.com/M0kY/react-form-stepper/compare/v1.2.7...v1.2.8) (2020-06-29)


### Bug Fixes

* specify tarball directory ([43130ec](https://github.com/M0kY/react-form-stepper/commit/43130ec9ed2ca2f4fda81e81bda6b6e6d98a2ad7))

## [1.2.7](https://github.com/M0kY/react-form-stepper/compare/v1.2.6...v1.2.7) (2020-06-29)


### Bug Fixes

* npm release ([7cf5bfa](https://github.com/M0kY/react-form-stepper/commit/7cf5bfafabea15f650b21da750c0a4c2065d61c7))

## [1.2.6](https://github.com/M0kY/react-form-stepper/compare/v1.2.5...v1.2.6) (2020-06-27)


### Bug Fixes

* indentation in workflow script ([2ab0c30](https://github.com/M0kY/react-form-stepper/commit/2ab0c30e91ab1bb1c7ffcae42ad08f8fbfa47908))
* release config and github action ([8e78834](https://github.com/M0kY/react-form-stepper/commit/8e7883478885a661455cb29d487fedcdd0f5a6ee))

## [1.2.5](https://github.com/M0kY/react-form-stepper/compare/v1.2.4...v1.2.5) (2020-05-17)


### Bug Fixes

* tsdx deprecation message ([4945c64](https://github.com/M0kY/react-form-stepper/commit/4945c64ab082b777be8374caa03eef5204fd18ed))

## [1.2.4](https://github.com/M0kY/react-form-stepper/compare/v1.2.3...v1.2.4) (2020-05-17)


### Bug Fixes


src/styles.ts
import { createUseStyles } from 'react-jss';
import { StepStyleProps, ConnectorStyleProps } from './types';

export const useStepperStyles = createUseStyles({
  StepperContainer: {
    display: 'flex',
    alignItems: 'flex-start',
    flexDirection: 'row',
    padding: 24,
    fontFamily: 'Arial, Helvetica, sans-serif',
  },
  StepContainer: {
    flex: 1,
    position: 'relative',
    paddingLeft: 8,
    paddingRight: 8,
  },
});

export const stepStyleDefaults = {
  activeBgColor: '#ed1d24',
  activeTextColor: '#ffffff',
  completedBgColor: '#a10308',
  completedTextColor: '#ffffff',
  inactiveBgColor: '#e0e0e0',
  inactiveTextColor: '#ffffff',
  size: '2em',
  circleFontSize: '1rem',
  labelFontSize: '0.875rem',
  borderRadius: '50%',
  fontWeight: 500,
};

export const useStepStyles = createUseStyles({
  StepMain: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  StepCircle: (props: StepStyleProps) => ({
    borderRadius: props.borderRadius,
    backgroundColor: props.inactiveBgColor,
    width: props.size,
    height: props.size,
    fontSize: '1em',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    '&.active': {
      backgroundColor: props.activeBgColor,
    },
    '&.completed': {
      backgroundColor: props.completedBgColor,
    },
  }),
  StepButton: {
    border: 'none',
    margin: 0,
    padding: 0,
    cursor: 'pointer',
  },
  StepCircleContent: (props: StepStyleProps) => ({
    color: props.inactiveTextColor,
    fontSize: props.circleFontSize,
    fontVariantNumeric: 'tabular-nums',
    '&.active': {
      color: props.activeTextColor,
    },
    '&.completed': {
      color: props.completedTextColor,
    },
  }),
  LabelContainer: (props: StepStyleProps) => ({
    width: '100%',
    fontSize: props.labelFontSize,
    fontWeight: 400,
    lineHeight: 1.4,
  }),
  Label: (props: StepStyleProps) => ({
    display: 'block',
    marginTop: 16,
    textAlign: 'center',
    fontWeight: props.fontWeight,
  }),
});

export const connectorStyleDefaults = {
  disabledColor: '#bdbdbd',
  activeColor: '#ed1d24',
  completedColor: '#a10308',
  size: 1,
  style: 'solid',
};

export const useConnectorStyles = createUseStyles({
  ConnectorContainer: (props: ConnectorStyleProps) => ({
    top: `calc((${props.stepSize} - ${props.size}) / 2)`,
    left: `calc(-50% + ${props.stepSize} - 8px)`,
    right: `calc(50% + ${props.stepSize} - 8px)`,
    position: 'absolute',
  }),
  Connector: (props: ConnectorStyleProps) => ({
    borderTopStyle: props.style,
    borderTopWidth: props.size,
    borderColor: props.disabledColor,
    display: 'block',
    '&.completed': {
      borderColor: props.completedColor,
    },
    '&.active': {
      borderColor: props.activeColor,
    },
  }),
});
